[DCS Patch 4/5] port dynamic channel selection from rdkb flow

itayx opened the pull request:
> The Dynamic Channel Selection as a feature can provide information on the supported channels which can be utilized by the management system to provide better availability and user experience.
#562

The Dynamic Channel Selection feature is divided into several pull request, this is the forth out of 5
### Dynamic channel selection message flow & actions
Changes in this PR include the following:
- agent: monitor_thread: proccess nl events
- agent: monitor_thread: handle channel scan events
- agent: monitor_thread: handle channel scan trigger CMDU
- controller & agent: master & slave: DCS channel scan
- controller: son_managment: DCS channel scan stubs
- controller: son_managment: DCS channel scan
- controller: bml: DCS channel scan setup
- controller: bml: DCS channel scan, set/get enable
- controller: bml: DCS channel scan, set/get params
- controller: bml: DCS channel scan, start single scan
- controller: bml: DCS channel scan, get results

**Before merging this PR please merge hotfix #695**

itayx committed:
controller & agent: master & slave: DCS channel scan

Requests, responses and notifications are propagated from
the agent to the controller and to the dynamic channel
selection task with no alteration to the data within

Create the task and add it to the task list on slave join
Handle channel scan trigger request and response
Handle channel scan triggered notification
Handle channel scan results notification
Handle channel scan finish notification
Handle channel scan abort notification

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: son_management: DCS channel scan stubs

This is a preparative commit for 'son_management: DCS channel scan'.

Add DCS task to includes
Add channel scan BML CMDUs to cases in "handle_bml_message"

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: son_management: DCS channel scan

The son_management's role is to handle incoming bml
requests, which means storing values in the DB,
launching an event to trigger the DCS scan in the task

Handle set/get enable request/response
Handle set/get params request/response
Handle get results request/response
Handle start single scan request/response

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: bml: DCS channel scan setup

The DCS channel scan feature is used via the BML APIs.

Ths is a preparative commit for all 'bml: DCS channel scan' commits.

Update the neighboring ap structure to fit new architecture
Add delayed response timeout to be used for results
Add cases to handle BML CMDUs responses
Add doxygen to the bml_internal APIs

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: bml: DCS channel scan, set/get enable

Get/Set the enable flag for the continuous DCS channel scan

Implement BML functionality
Add bml_internal API
Implement response CMDU handling

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: bml: DCS channel scan, set/get params

Get/Set the scan parameters for the continuous DCS channel scan

Implement BML functionality
Add bml_internal API
Add response promise handling
Implement response CMDU handling

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: bml: DCS channel scan, start single scan

Start the single (one time) channel scan with parameters

Implement BML functionality
Add bml_internal API
Implement response CMDU handling

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: bml: DCS channel scan, get results

Get the scan results for both the single and continuous DCS channel scan

Implement BML functionality
Add bml_internal API
Add response promise handling
Implement response CMDU handling
Add translation function from BML struct to CMDU struct

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
agent: monitor_thread: process nl events

The dynamic channel selection feature uses the NL and dwpal
apis to trigger a neighboring scan and uses the results
received for analytical data.

Open the NL sockets and start listening for events
Process the NL events and add handling stubs

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
agent: monitor_thread: handle channel scan events

The monitor thread receives channel scan events sent from
the wlan hal dwpal and handles them accordingly, by
propagating them to the agent

Add handling for Channel_Scan_Triggered
Add handling for Channel_Scan_New_Results_Ready
Add handling for Channel_Scan_Dump_Result
Add handling for Channel_Scan_Finished
Add handling for Channel_Scan_Abort

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
agent: monitor_thread: handle channel scan trigger CMDU

The agent can request the monitor thread to trigger a
channel scan

Receive the channel scan trigger CMDU
Extract parameters and print descriptive log
Trigger the channel scan with parameters

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -79,7 +79,7 @@ void monitor_thread::stop_monitor_thread()
 
         LOG(ERROR) << "disconnecting monitor_thread sockets";
-        if (mon_hal_ext_events || slave_socket)
+        if (mon_hal_ext_events || mon_hal_nl_events || slave_socket)
```
We always use `{}` with `if` statements (even one-liners)

vitalybu replied:
What is the purpose of this `if` statement anyway?

itayx replied:
To log that at least one of the sockets is about to be disconnected
removed the first "if" but added LOG(DEBUG) << "stopping <socket name>" for each individual socket

tomereli replied:
> We always use `{}` with `if` statements (even one-liners)

You are right, unfortunately, according to the majority :)
https://doodle.com/poll/wwyfaymx84chhe3k

tomereli replied:
Also - please use `LOG_IF` instead of `if` when only printing logs based on condition.

itayx replied:
Changed it to `LOG_IF`

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -101,6 +101,13 @@ void monitor_thread::stop_monitor_thread()
+            remove_socket(mon_hal_nl_events);
+            delete mon_hal_nl_events;
+            mon_hal_nl_events = nullptr;
+            LOG(ERROR) << "nl socket remove";
```
Remove this log

itayx replied:
removed

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -262,6 +274,21 @@ void monitor_thread::after_select(bool timeout)
+            if (nl_events_fd > 0) {
+                mon_hal_nl_events = new Socket(nl_events_fd);
+                add_socket(mon_hal_nl_events);
+                LOG(DEBUG) << "nl socket created ";
```
Maybe it's a good idea to print the FD of the newly created socket... Can be useful in future logging

itayx replied:
Added a print

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -262,6 +274,21 @@ void monitor_thread::after_select(bool timeout)
+                mon_hal_nl_events = new Socket(nl_events_fd);
+                add_socket(mon_hal_nl_events);
+                LOG(DEBUG) << "nl socket created ";
+            } else if (nl_events_fd == 0) {
```
I don't think it's a valid case.. Can be removed and handled by the `else` statement

itayx replied:
@kantera800
What do you think?
`get_nl_events_fd()` description says  it returns -1 on error but has no mention of other cases

kantera800 replied:
Let's remove it

itayx replied:
Removed & moved the `mon_hal_nl_events = nullptr` to the (now only) `else` statement

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -346,6 +373,26 @@ void monitor_thread::after_select(bool timeout)
         }
 
+        // Process nl events
+        if (mon_hal_nl_events) {
```
`mon_hal_nl_events` is actually a Socket instance, so perhaps `mon_hal_nl_socket` is a better name

itayx replied:
since `mon_hal_ext_events` uses the same naming convention, I don't think we should refactor the naming here unless we change that one as well

tomereli replied:
You're both right, but I guess we should keep the naming convention change to a separate PR.

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -346,6 +373,26 @@ void monitor_thread::after_select(bool timeout)
+                    stop_monitor_thread();
+                    return;
+                }
+                LOG(DEBUG) << "proccess nl events finished ";
```
Remove

itayx replied:
removed

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -346,6 +373,26 @@ void monitor_thread::after_select(bool timeout)
+            thread_last_error_code = MONITOR_THREAD_ERROR_NL_ATTACH_FAIL;
+            stop_monitor_thread();
+            return;
+        }
```
Can this ever happen?
`stop_monitor_thread()` is called if the FD is not valid..

itayx replied:
Removed

kantera800 replied:
in case platform dosnt support netlink (dummy mode for example) we shouldn't call stop_monitor_thread() in cased FD is invalid.
@itayx  didn't you add a fixup that removes it ?

itayx replied:
yeah, I commented saying I did remove it

kantera800 replied:
was not clear :)

vitalybu commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3323,6 +3324,103 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+
+        dynamic_channel_selection_task::sScanEvent new_event;
+        new_event.radio_mac = radio_mac;
+        tasks.push_event(database.get_dynamic_channel_selection_task_id(radio_mac),
```
What if there's no DCS task for this MAC?

itayx replied:
Added a check and an error log in case it doesn't find one

vitalybu commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3323,6 +3324,103 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+        auto radio_mac = notification->radio_mac();
+
+        //send results to dcs task
+        if (database.get_channel_scan_in_progress(radio_mac, false) ||
+            database.get_channel_scan_in_progress(radio_mac, true)) {
```
Is it possible to change the API to allow checking both conditions in a single call?

itayx replied:
I could add another api called `get_any_channel_scan_in_progress(const sMacAddr&)` but I think that would be redundant as this will be the only place it will be used

tomereli replied:
So at least provide an explanation in the form of a comment before the if condition - when I read this code I don't understand how `false` and `true` differ.

itayx replied:
Added a comment explaining it

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -147,7 +148,7 @@ bool monitor_thread::socket_disconnected(Socket *sd)
         stop_monitor_thread();
         return false;
-    }
+     }
```
Remove the extra space

itayx replied:
removed

vitalybu commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1153,6 +1154,52 @@ bool monitor_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+        auto channel_pool_vector =
+            std::vector<unsigned int>(channel_pool, channel_pool + channel_pool_size);
+        std::string channels;
+        for (int8_t index = 0; index < channel_pool_size; index++) {
```
Use regular `int` instead of `int8_t`
Also, add a newline separation between variable declaration and code segments

vitalybu replied:
This loop is for debug purposes only? If so, please add a relevant comment

itayx replied:
Changed the `uint8_t` to `int`.
Added a comment explaining this loop is only for the print.
And made the whole loop thinner by using a ternary operator instead of the if statement
`channels += ((index != 0) ? "," : "") + std::to_string(int(channel_pool[index]));`

vitalybu commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1803,26 +1803,355 @@ void son_management::handle_bml_message(Socket *sd,
+
+        // Get information from request
+        auto radio_mac      = request->radio_mac();
+        auto is_single_scan = request->scan_mode() == 1;
```
There's no constant or enum for the scan modes?
Are there any more modes or is it just 0 and 1?

itayx replied:
just 0 & 1 since we don't send booleans over CMDU

tomereli replied:
So define an enum in the yaml and use it instead

tomereli replied:
nitpick anyway..

vitalybu commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1803,26 +1803,355 @@ void son_management::handle_bml_message(Socket *sd,
+        //Results avaliablity check
+        if (scan_results_size == 0) {
+            LOG(DEBUG) << "no scan results are available";
+            //op_error_code = eChannelScanOpErrCode::CHANNEL_SCAN_OP_SCAN_RESULTS_EMPTY;
```
Remove unused code

itayx replied:
Moved the whole segment to fix an issue with sending CHANNEL_SCAN_RESULTS_EMPTY

vitalybu reviewed (dismissed):
Reviewed till: `controller: bml: DCS channel scan setup`

itayx commented:
This PR is dependent on hotfix #695

arnout commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[DCS Patch 4/5]Port dynamic channel selection](https://trello.com/c/pV5WJCd8/78-dcs-patch-4-5port-dynamic-channel-selection)

tomereli commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -135,6 +142,11 @@ bool monitor_thread::socket_disconnected(Socket *sd)
         return false;
+    } else if (mon_hal_nl_events && (sd == mon_hal_nl_events)) {
+        LOG(ERROR) << "mon_hal_nl_events socket disconnected!";
+        thread_last_error_code = MONITOR_THREAD_ERROR_HAL_DISCONNECTED;
```
This is incorrect, should be `thread_last_error_code = MONITOR_THREAD_ERROR_NL_EVENTS_SOCKET_DISCONNECTED`;

itayx replied:
Changed

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -2788,6 +2812,105 @@ bool slave_thread::handle_cmdu_monitor_message(Socket *sd,
+        }
+
+        notification_out->scan_results() = notification_in->scan_results();
+        notification_out->is_dump()      = notification_in->is_dump();
```
What is the `is_dump()` parameter? Can it be part of the `scan_results()`?

itayx replied:
No, `is_dump()` is a flag that is only used in the "Get Scan Results" messages
while `scan_results()` is a structure called `sChannelScanResults` which is used throughout the code

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3323,6 +3324,103 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+        LOG(TRACE) << "DCS_task, sending SCAN_TRIGGERED for mac " << hostap_mac;
+        auto notification =
+            beerocks_header->addClass<beerocks_message::cACTION_CONTROL_CHANNEL_SCAN_TRIGGERED_NOTIFICATION>();
+        if (notification == nullptr) {
```
`if (!notification)` to avoid KW issues.
Have you run KW on this PR? Please do so.

itayx replied:
Changed

rmelotte commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3323,6 +3324,103 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+
+        tasks.push_event(database.get_dynamic_channel_selection_task_id(radio_mac),
+                         (int)dynamic_channel_selection_task::eEvent::SCAN_ABORTED,
+                         (void *)&new_event);
```
No C-style cast, use a more appropriate type of cast (probably a static_cast).

rmelotte commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3323,6 +3324,103 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+        LOG(DEBUG) << "DCS_task, sending SCAN_ABORTED for mac " << hostap_mac;
+        auto notification =
+            beerocks_header->addClass<beerocks_message::cACTION_CONTROL_CHANNEL_SCAN_ABORT_NOTIFICATION>();
+        if (notification == nullptr) {
```
this should be `if (!notification) {`.

itayx replied:
Fixed

rmelotte commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3323,6 +3324,103 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+
+        tasks.push_event(database.get_dynamic_channel_selection_task_id(radio_mac),
+                         (int)dynamic_channel_selection_task::eEvent::SCAN_FINISHED,
+                         (void *)&new_event);
```
avoid c-style casts.

itayx replied:
Fixed

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3323,6 +3324,103 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+            return false;
+        }
+
+        auto radio_mac = notification->radio_mac();
```
Why do you need an extra variable definition? Just use:
```
dynamic_channel_selection_task::sScanEvent new_event;
new_event.radio_mac = notification->radio_mac();
tasks.push_event(database.get_dynamic_channel_selection_task_id(notification->radio_mac()),
     (int)dynamic_channel_selection_task::eEvent::SCAN_TRIGGERED,
     (void *)&new_event);
```
(nitpick)

itayx replied:
Fixed

rmelotte commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1106,6 +1152,52 @@ bool monitor_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+        //debug print incoming information:
+        LOG(DEBUG) << std::endl
+                   << "scan_params:" << std::endl
+                   << "radio_mac=" << network_utils::mac_to_string(radio_mac) << std::endl
```
No need to use mac_to_string, the << operator should already be defined for `sMacAddr`.

itayx replied:
Changed

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -3323,6 +3324,103 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+        }
+
+        auto radio_mac = notification->radio_mac();
+
```
remove extra line break (here and everywhere else)

tomereli commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1800,6 +1801,34 @@ void son_management::handle_bml_message(Socket *sd,
         break;
     }
+    case beerocks_message::ACTION_BML_CHANNEL_SCAN_SET_CONTINUOUS_PARAMS_REQUEST: {
+        LOG(TRACE) << "ACTION_BML_CHANNEL_SCAN_SET_CONTINUOUS_PARAMS_REQUEST";
```
Remove extra space

itayx replied:
It was removed in the next commit

itayx replied:
It was removed in the next commit

tomereli commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1417,13 +1417,50 @@ bool monitor_thread::hal_event_handler(bwl::base_wlan_hal::hal_event_ptr_t event
+        }
+
+        // If event == Channel_Scan_New_Results_Ready do nothing since is_dump's default is 0
+        if (event == Event::Channel_Scan_Dump_Result) {
```
I think a one-liner is enough here: `notification->is_dump() = (event == Event::Channel_Scan_Dump_Result)` (nitpick)

itayx replied:
This is a preparation for the 5th PR
When `event == Event::Channel_Scan_Dump_Result` not only do we set the `is_dump() = 1`
but also copy the data received with the event to the notification

tomereli commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -601,7 +601,7 @@ int bml_get_dcs_continuous_scan_params(BML_CTX ctx, const char *radio_mac, int *
  */
 int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
-                             struct BML_DCS_NEIGHBOR_AP **output_results,
+                             struct BML_NEIGHBOR_AP **output_results,
```
I now understand the change - it replaces the results structure as we discussed.
This is a good example where a commit message should provide much more context, so the reviewer can understand why this change was done.

All the commit messages in this PR (and the other PRs) are in the format of "Do this, do that" - this is not in par with the prplMesh contribution guide.

Therefore, please change *all* commit messages in this PR (and future PRs) to abide by the prplMesh standard defined in https://github.com/prplfoundation/prplMesh/blob/master/CONTRIBUTING.md#commits.

tomereli replied:
?

rmelotte commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -101,6 +102,13 @@ void monitor_thread::stop_monitor_thread()
 
+        if (mon_hal_nl_events) {
+            remove_socket(mon_hal_nl_events);
+            delete mon_hal_nl_events;
```
By looking at the rest of the file I wonder why we have to resort to explicitly calling delete here, though that's not related to this PR.

itayx replied:
@kantera800
I'm not sure myself if this is needed

kantera800 replied:
from looking at the code remove_socket just isn't strong enough

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -27,8 +27,9 @@ INITIALIZE_EASYLOGGINGPP
-#define RESPONSE_TIMEOUT (5000) // 5 seconds
+#define SELECT_TIMEOUT (500)             // 500 milliseconds
+#define RESPONSE_TIMEOUT (5000)          // 5 seconds
+#define DELAYED_RESPONSE_TIMEOUT (30000) // 30 seconds
```
What is this used for? Please add a more elaborate comment (do this for all the timeouts while you're at it).

itayx replied:
They are used as timeouts for promises set on responses to requests
and are mostly used like the following: `<promise>.wait_for(<timeout>)`

I've added a comment explaining this

rmelotte commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1106,6 +1152,52 @@ bool monitor_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+            if (index != 0) {
+                channels += ",";
+            }
+            channels += std::to_string(int(channel_pool[index]));
```
Why does it need to be converted to an int before making the std::string ?

itayx replied:
Fixed

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1219,6 +1219,53 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
 
+int bml_internal::set_dcs_continuous_scan_enable(const std::string &mac, int enable)
+{
+    //CMDU message
```
The comment is meaningless - all are CMDU messages.

itayx replied:
Removed all `//CMDU message`

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1219,6 +1219,53 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+        return (-BML_RET_OP_FAILED);
+    }
+
+    request->radio_mac() = network_utils::mac_from_string(mac);
```
Please, no `mac_from_string` - use sMacAddr in the YAML instead of string unless you have a very good reason not to.

itayx replied:
The feature must use the mac_from_string at least once since the bml gets the mac address as a string buffer (char[]) in the bml functions
So I moved the mac_from_string to the bml

tomereli replied:
OK

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1219,6 +1219,53 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+        return (-BML_RET_OP_FAILED);
+    }
+
+    if (result > 0) {
```
Can result be negative?

itayx replied:
No, changed it to `!= int(eChannelScanOpErrCode::CHANNEL_SCAN_OP_SUCCESS)`

rmelotte commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1106,6 +1152,52 @@ bool monitor_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+        auto channel_pool_vector =
+            std::vector<unsigned int>(channel_pool, channel_pool + channel_pool_size);
+        std::string channels;
+        for (int8_t index = 0; index < channel_pool_size; index++) {
```
`channel_pool_size` is a `uint8_t`, `index` should probably also be a uint8_t.

Unrelated to this PR:
channel_pool_size should also probably be of a size type since it's a size.

itayx replied:
Changed the for loop to
`for (int index = 0; index < int(channel_pool_size); index++) {`
that way it is not dependent on what type is the `channel_pool_size`

And regarding the `channel_pool_size`, it needs to be a basic type for the CMDUs

rmelotte commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -564,8 +564,8 @@ int bml_set_dcs_continuous_scan_enable(BML_CTX ctx, const char *radio_mac, int e
 
-    // TODO: call suitable bml api
-    return BML_RET_OP_FAILED;
+    bml_internal *pBML = (bml_internal *)ctx;
```
Don't use c-style casts (see other comments).

itayx replied:
Changed

rmelotte commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -575,8 +575,8 @@ int bml_get_dcs_continuous_scan_enable(BML_CTX ctx, const char *radio_mac, int *
 
-    // TODO: call suitable bml api
-    return BML_RET_OP_FAILED;
+    bml_internal *pBML = (bml_internal *)ctx;
```
Don't use c-style casts (see other comments).

itayx replied:
Changed

rmelotte commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -588,8 +588,9 @@ int bml_set_dcs_continuous_scan_params(BML_CTX ctx, const char *radio_mac, int d
 
-    // TODO: call suitable bml api
-    return BML_RET_OP_FAILED;
+    bml_internal *pBML = (bml_internal *)ctx;
```
Don't use c-style casts (see other comments).

itayx replied:
Changed

rmelotte commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -615,8 +617,9 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
 
-    // TODO: call suitable bml api
-    return BML_RET_OP_FAILED;
+    bml_internal *pBML = (bml_internal *)ctx;
```
Don't use c-style casts (see other comments).

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+        return (-BML_RET_OP_FAILED);
+    }
+
+    request->radio_mac() = network_utils::mac_from_string(mac);
```
Don't use string for mac addresses, use sMacAddr. It would avoid the need to convert it back here.

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+    return send_bml_cmdu(*output_enable, request->get_action_op());
+}
+
+int bml_internal::set_dcs_continuous_scan_params(const std::string &mac, int dwell_time,
```
Don't use a string for `mac` if possible (use sMacAddr).

itayx replied:
[See](https://github.com/prplfoundation/prplMesh/pull/694/commits/3104d970c25b24adcf45be9e062c5d8b5a68a96e#r369162085)

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+    }
+
+    // If the socket is not valid, attempt to re-establish the connection
+    if (m_sockMaster == nullptr) {
```
`if (! m_sockMaster) {`

itayx replied:
Fixed

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+    request->params().interval_time_sec = interval_time;
+    request->params().channel_pool_size = channel_pool_size;
+    if (channel_pool_size > 0 && channel_pool_size <= BML_CHANNEL_SCAN_MAX_CHANNEL_POOL_SIZE &&
+        channel_pool != nullptr) {
```
compare pointers with (! pointer), not `==nullptr`.

itayx replied:
Changed

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+                                                 int *output_channel_pool_size)
+{
+    if (output_dwell_time == nullptr && output_interval_time == nullptr &&
+        output_channel_pool_size == nullptr) {
```
compare pointers with (! pointer), not `==nullptr`.

itayx replied:
changed

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+        return (-BML_RET_INVALID_DATA);
+    }
+
+    if (m_sockPlatform == nullptr && !connect_to_platform()) {
```
compare pointers with (! pointer), not `==nullptr`.

itayx replied:
changed

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+        return (-BML_RET_OP_FAILED);
+    }
+
+    request->radio_mac() = network_utils::mac_from_string(mac);
```
mac should already be a sMacAddr so that you don't need to convert it.

itayx replied:
[See](https://github.com/prplfoundation/prplMesh/pull/694/commits/3104d970c25b24adcf45be9e062c5d8b5a68a96e#r369162085)

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+    // Clear the promise holder
+    m_prmChannelScanParamsGet = nullptr;
+
+    if (output_dwell_time != nullptr) {
```
compare pointers with (! pointer), not `==nullptr`.

itayx replied:
changed

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+    if (output_dwell_time != nullptr) {
+        *output_dwell_time = ScanParams.dwell_time_ms;
+    }
+    if (output_interval_time != nullptr) {
```
compare pointers with (! pointer), not `==nullptr`.

itayx replied:
changed

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+    if (output_interval_time != nullptr) {
+        *output_interval_time = ScanParams.interval_time_sec;
+    }
+    if (output_channel_pool_size != nullptr && output_channel_pool != nullptr) {
```
compare pointers with (! pointer), not `==nullptr`.

itayx replied:
changed

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1206,6 +1251,309 @@ int bml_internal::bml_get_vap_list_credentials(BML_VAP_INFO *vaps, uint8_t &vaps
+                                       const unsigned int max_results_size,
+                                       uint8_t *output_result_status, bool is_single_scan)
+{
+    if (m_sockPlatform == nullptr && !connect_to_platform()) {
```
compare pointers with (! pointer), not `==nullptr`.

itayx replied:
Fixed

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.h:
```diff
@@ -152,6 +152,9 @@ class bml_internal : public beerocks::socket_thread {
 
+    //set and get channel scan enable flag
+    int set_dcs_continuous_scan_enable(const std::string &mac, int enable);
+    int get_dcs_continuous_scan_enable(const std::string &mac, int *output_enable);
```
Please change to `int get_dcs_continuous_scan_enable(const std::string &mac, int *enable);` and add doxygen marking `int *enable` as `[OUT]` in the doxygen comment.

vitalybu replied:
Why not reference?

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.h:
```diff
@@ -152,6 +152,25 @@ class bml_internal : public beerocks::socket_thread {
+    int get_dcs_continuous_scan_enable(const std::string &mac, int *output_enable);
+
+    //set and get channel scan params
+    int set_dcs_continuous_scan_params(const std::string &mac, int dwell_time, int interval_time,
```
Don't use std::string for mac, use sMacAddr.

itayx replied:
[See](https://github.com/prplfoundation/prplMesh/pull/694/commits/3104d970c25b24adcf45be9e062c5d8b5a68a96e#r369162085)

rmelotte commented on controller/src/beerocks/bml/internal/bml_internal.h:
```diff
@@ -152,6 +152,25 @@ class bml_internal : public beerocks::socket_thread {
+    //set and get channel scan params
+    int set_dcs_continuous_scan_params(const std::string &mac, int dwell_time, int interval_time,
+                                       unsigned int *channel_pool, int channel_pool_size);
+    int get_dcs_continuous_scan_params(const std::string &mac, int *output_dwell_time,
```
Don't use std::string for mac, use sMacAddr.

itayx replied:
[See](https://github.com/prplfoundation/prplMesh/pull/694/commits/3104d970c25b24adcf45be9e062c5d8b5a68a96e#r369162085)

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+    }
+
+    // If the socket is not valid, attempt to re-establish the connection
+    if (m_sockMaster == nullptr) {
```
`if (!m_sockMaster)`

itayx replied:
Fixed

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+        }
+    }
+
+    //CMDU message
```
meaningless comment

itayx replied:
removed

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+    request->params().interval_time_sec = interval_time;
+    request->params().channel_pool_size = channel_pool_size;
+    if (channel_pool_size > 0 && channel_pool_size <= BML_CHANNEL_SCAN_MAX_CHANNEL_POOL_SIZE &&
+        channel_pool != nullptr) {
```
`!channel_pool`

itayx replied:
Changed

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+        return (-BML_RET_OP_FAILED);
+    }
+
+    if (result > 0) {
```
Again - can result be negative? what is considered as success? shouldn't this be simple boolean?

itayx replied:
changed to check the result against the corresponding enum

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+        return (-BML_RET_OP_FAILED);
+    }
+
+    request->radio_mac() = network_utils::mac_from_string(mac);
```
no mac_from_string please

itayx replied:
[See](https://github.com/prplfoundation/prplMesh/pull/694/commits/3104d970c25b24adcf45be9e062c5d8b5a68a96e#r369162085)

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+    // Clear the promise holder
+    m_prmChannelScanParamsGet = nullptr;
+
+    if (output_dwell_time != nullptr) {
```
`if (output_dwell_time)`

itayx replied:
changed

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+    if (output_dwell_time != nullptr) {
+        *output_dwell_time = ScanParams.dwell_time_ms;
+    }
+    if (output_interval_time != nullptr) {
```
same here

itayx replied:
changed

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+    if (output_interval_time != nullptr) {
+        *output_interval_time = ScanParams.interval_time_sec;
+    }
+    if (output_channel_pool_size != nullptr && output_channel_pool != nullptr) {
```
same here

itayx replied:
changed

rmelotte reviewed (commented):
Did a quick code review, lot of the comments are the same. Note that I only blindly reviewed the _code_ itself, not the overall changes, because I don't know enough about it to review it. For this reason I'm only adding this as a comment, not approving nor requesting changes (please take the comments into account though).

Most were:
- There were many uses of `ptr == nullptr` to compare pointers, you should use `! ptr` directly. There were a lot of these, I didn't add a comment for all of them.

- Please also remove all C style-casts and use a more appropriate one (probably static_cast in all cases).

- Also don't use std::string for mac addresses, use sMacAddr instead.

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1401,6 +1401,44 @@ int bml_internal::get_dcs_continuous_scan_params(const std::string &mac, int *ou
+                                        int channel_pool_size, unsigned int *channel_pool)
+{
+    LOG(DEBUG) << "start_single_channel_scan";
+    //CMDU message
```
remove comment

itayx replied:
Removed all `//CMDU message`

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -64,6 +64,38 @@ static void config_logger(const std::string log_file = std::string())
 
 #endif // BEEROCKS_DEBUG
 
+static void translate_channel_scan_results(const beerocks_message::sChannelScanResults &res_in,
```
Why not change to return `bool` if it can fail?

itayx replied:
All it does is copy so it's can't really fail unless it fails to allocate memory

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1401,6 +1433,89 @@ int bml_internal::get_dcs_continuous_scan_params(const std::string &mac, int *ou
+                                       const unsigned int max_results_size,
+                                       uint8_t *output_result_status, bool is_single_scan)
+{
+    if (m_sockPlatform == nullptr && !connect_to_platform()) {
```
Also, please separate to 2 separate if conditions.

itayx replied:
Changed

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1401,6 +1433,89 @@ int bml_internal::get_dcs_continuous_scan_params(const std::string &mac, int *ou
+    m_scan_results_maxsize = &scan_results_maxsize;
+    m_scan_results_status  = output_result_status;
+
+    //CMDU message
```
remove comment

itayx replied:
Removed all `//CMDU message`

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1401,6 +1433,89 @@ int bml_internal::get_dcs_continuous_scan_params(const std::string &mac, int *ou
+    //output_results_size will be set to the number of actual returning results
+    *output_results_size = 0;
+    for (auto res : scan_results) {
+        auto out = &((*output_results)[*output_results_size]);
```
`auto &`

itayx replied:
Fixed

tomereli replied:
I meant to pass the output result as a reference.

Change the declaration of `translate_channel_scan_results` to accept `BML_NEIGHBOR_AP &res_out` instead of `BML_NEIGHBOR_AP *res_out` and set `auto &out = (*output_results)[*output_results_size];`

itayx replied:
Oh.. my bad
Fixed and added a comment explaining the whole `(*output_results)[*output_results_size]` deal so it would be more understandable

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.h:
```diff
@@ -240,15 +246,19 @@ class bml_internal : public beerocks::socket_thread {
+    std::list<beerocks_message::sChannelScanResults> *m_scan_results = nullptr;
+    uint8_t *m_scan_results_status                                   = nullptr;
+    uint32_t *m_scan_results_size                                    = nullptr;
+    uint32_t *m_scan_results_maxsize                                 = nullptr;
```
Please add a comment explaining each member.

itayx replied:
Added

tomereli commented:
Channel selection test fails with this pr

tomereli commented:
Please fix the channel selection test, then rebase squash and we'll do the followup review.

kantera800 commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -346,6 +373,24 @@ void monitor_thread::after_select(bool timeout)
+                }
+                clear_ready(mon_hal_nl_events);
+            }
+        } else {
```
@itay please  remove this  else completely.
We are entering here each after_select  no need for this print at all.

itayx replied:
removed

vitalybu commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1266,6 +1266,141 @@ int bml_internal::get_dcs_continuous_scan_enable(const std::string &mac, int *ou
+                                                 int interval_time, unsigned int *channel_pool,
+                                                 int channel_pool_size)
+{
+    if (dwell_time == BML_CHANNEL_SCAN_INVALID_PARAM &&
+        interval_time == BML_CHANNEL_SCAN_INVALID_PARAM &&
+        channel_pool_size == BML_CHANNEL_SCAN_INVALID_PARAM) {
+        LOG(ERROR) << "Function is called, but no data is being set!";
+        return (-BML_RET_INVALID_DATA);
+    }
```
What if only one of the parameters has invalid value?
Shouldn't this statement use OR conditions?

itayx replied:
No, this handles the case if no parameters is being requested by the caller
If all params are invalid it means that there is no need to continue with the function
yet, even if one is valid  it will still continue but return only that value

tomereli replied:
Any reason this conversation is unresolved?

vitalybu commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -64,6 +64,38 @@ static void config_logger(const std::string log_file = std::string())
+        return;
+    }
+
+    string_utils::copy_string(res_out->ap_SSID, res_in.ssid, 36);
```
Use a constant for the SSID size

itayx replied:
Changed

tomereli commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -564,8 +564,9 @@ int bml_set_dcs_continuous_scan_enable(BML_CTX ctx, const char *radio_mac, int e
+        network_utils::mac_from_string(std::string(radio_mac)), enable);
 }
 
 int bml_get_dcs_continuous_scan_enable(BML_CTX ctx, const char *radio_mac, int *output_enable)
```
Change to `int bml_get_dcs_continuous_scan_enable(BML_CTX ctx, const char *radio_mac, int *enable)`

Or even better as Vitaly suggested - why not use reference?

itayx replied:
In the BML I can't use a reference, since I'm not sure how well it will be imported in C
I can change the `bml_internal` to use a reference instead

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.h:
```diff
@@ -152,6 +152,9 @@ class bml_internal : public beerocks::socket_thread {
 
+    //set and get channel scan enable flag
+    int set_dcs_continuous_scan_enable(const sMacAddr &mac, int enable);
+    int get_dcs_continuous_scan_enable(const sMacAddr &mac, int *enable);
```
I prefer `int bml_get_dcs_continuous_scan_enable(BML_CTX ctx, const char *radio_mac, int &enable)`

itayx replied:
the BML cannot use a reference since it's used as a C API
but the bml_internal uses reference

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -67,6 +67,39 @@ static void config_logger(const std::string log_file = std::string())
 
 #endif // BEEROCKS_DEBUG
 
+static void translate_channel_scan_results(const beerocks_message::sChannelScanResults &res_in,
```
Change to `static void translate_channel_scan_results(const beerocks_message::sChannelScanResults &res_in, BML_NEIGHBOR_AP &res_out)`

Please always use reference whenever possible.

itayx replied:
Changed

tomereli reviewed (changes requested):
Only small comments only but since the commit messages where not changed yet I'm not approving.
Please fix the new comments, update commit messages appropriately, rebase + squash and ask for a final re-review.

itayx committed:
hotfix: fix clear_channel_scan_results()

was clearing config::channel_pool instead scan_results

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
hotfix: spamming logs removed and fixed incorrect logs

Remove the "current state" and "next state" logs
Move the log and compare functions to the cpp file
Add log on fsm_move_state

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>
